File: python/url_parser/implementation/url_parser.py
from dataclasses import dataclass


@dataclass
class ParsedURL:
    scheme: str
    userinfo: str | None
    host: str
    port: int | None
    path: str
    query: str | None
    fragment: str | None

    def reconstruct(self) -> str:
        parts: list[str] = []
        parts.append(f"{self.scheme}://")
        if self.userinfo is not None:
            parts.append(f"{self.userinfo}@")
        if ":" in self.host:
            parts.append(f"[{self.host}]")
        else:
            parts.append(self.host)
        default_port = _default_ports.get(self.scheme)
        if self.port is not None and self.port != default_port:
            parts.append(f":{self.port}")
        parts.append(self.path)
        if self.query is not None:
            parts.append(f"?{self.query}")
        if self.fragment is not None:
            parts.append(f"#{self.fragment}")
        return "".join(parts)

    def validate(self) -> bool:
        if not self.scheme:
            return False
        if not self.host:
            return False
        if self.port is not None and not (0 < self.port <= 65535):
            return False
        return True


_default_ports: dict[str, int] = {
    "http": 80,
    "https": 443,
    "ftp": 21,
    "ssh": 22,
    "smtp": 25,
    "ws": 80,
    "wss": 443,
}


class URLParser:
    @staticmethod
    def parse(url: str) -> ParsedURL:
        if not url:
            raise ValueError("Empty URL")

        fragment: str | None = None
        if "#" in url:
            url, fragment = url.rsplit("#", 1)
            if fragment == "":
                fragment = None

        query: str | None = None
        if "?" in url:
            url, query = url.rsplit("?", 1)
            if query == "":
                query = None

        scheme = ""
        userinfo: str | None = None
        host = ""
        port: int | None = None
        path = ""

        if "://" in url:
            scheme, rest = url.split("://", 1)
            if not scheme:
                raise ValueError("Missing scheme")
            if "/" in rest:
                authority, path = rest.split("/", 1)
                path = "/" + path
            else:
                authority = rest
                path = ""

            if "@" in authority:
                userinfo, authority = authority.rsplit("@", 1)
                if not userinfo:
                    userinfo = None

            if authority.startswith("["):
                if "]" not in authority:
                    raise ValueError("Unclosed IPv6 bracket")
                host, _, maybe_port = authority[1:].partition("]")
                if maybe_port:
                    if not maybe_port.startswith(":"):
                        raise ValueError("Invalid IPv6 syntax")
                    port_str = maybe_port[1:]
                    port = URLParser._parse_port(port_str) if port_str else None
                else:
                    port = _default_ports.get(scheme)
            else:
                if ":" in authority:
                    host, port_str = authority.rsplit(":", 1)
                    port = URLParser._parse_port(port_str)
                else:
                    host = authority
                    port = _default_ports.get(scheme)

            if not host:
                raise ValueError("Missing host")

        else:
            path = url

        return ParsedURL(
            scheme=scheme,
            userinfo=userinfo,
            host=host,
            port=port,
            path=path or "/",
            query=query,
            fragment=fragment,
        )

    @staticmethod
    def _parse_port(port_str: str) -> int:
        try:
            port = int(port_str)
        except ValueError:
            raise ValueError(f"Invalid port: {port_str}")
        if not (0 < port <= 65535):
            raise ValueError(f"Port out of range: {port}")
        return port

File: java/url_parser/src/main/java/com/qe/test/parser/UrlParser.java
package com.qe.test.parser;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Pattern;

public class UrlParser {
    private static final Pattern URL_PATTERN = Pattern.compile(
            "^([a-zA-Z][a-zA-Z0-9+\\-.]*):" +                    // scheme
            "(?://(?:(.*?)@)?" +                                   // userinfo@
            "(\\[[a-fA-F0-9:]+\\]|[^:/?#]*)" +                    // host (IPv6 or regular)
            "(?::(\\d*))?" +                                       // port
            ")?(/[^?#]*)?" +                                       // path
            "(?:\\?([^#]*))?" +                                    // query
            "(?:#(.*))?$"                                          // fragment
    );

    private static final Pattern QUERY_PARAM_PATTERN = Pattern.compile("([^&=]+)=([^&]*)");

    private final String scheme;
    private final String userInfo;
    private final String host;
    private final int port;
    private final String path;
    private final String query;
    private final String fragment;
    private final Map<String, String> queryParams;
    private final String original;

    private UrlParser(Builder builder) {
        this.scheme = builder.scheme;
        this.userInfo = builder.userInfo;
        this.host = builder.host;
        this.port = builder.port;
        this.path = builder.path;
        this.query = builder.query;
        this.fragment = builder.fragment;
        this.queryParams = Collections.unmodifiableMap(builder.queryParams);
        this.original = builder.original;
    }

    public static UrlParser parse(String url) {
        if (url == null) {
            throw new NullPointerException("url must not be null");
        }
        var matcher = URL_PATTERN.matcher(url);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("Malformed URL: " + url);
        }

        Builder builder = new Builder(url);
        builder.scheme = matcher.group(1).toLowerCase();
        builder.userInfo = matcher.group(2);

        String hostStr = matcher.group(3);
        if (hostStr != null) {
            if (hostStr.startsWith("[") && hostStr.endsWith("]")) {
                builder.host = hostStr.substring(1, hostStr.length() - 1);
            } else {
                builder.host = hostStr;
            }
        }

        String portStr = matcher.group(4);
        if (portStr != null && !portStr.isEmpty()) {
            try {
                builder.port = Integer.parseInt(portStr);
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("Invalid port: " + portStr, e);
            }
        } else {
            builder.port = getDefaultPort(builder.scheme);
        }

        String pathStr = matcher.group(5);
        builder.path = pathStr != null ? pathStr : "";

        builder.query = matcher.group(6);
        if (builder.query != null && !builder.query.isEmpty()) {
            builder.queryParams = parseQueryParams(builder.query);
        }

        builder.fragment = matcher.group(7);

        return builder.build();
    }

    private static Map<String, String> parseQueryParams(String query) {
        Map<String, String> params = new LinkedHashMap<>();
        var matcher = QUERY_PARAM_PATTERN.matcher(query);
        while (matcher.find()) {
            params.put(decodePercent(matcher.group(1)), decodePercent(matcher.group(2)));
        }
        return params;
    }

    static String decodePercent(String s) {
        try {
            return java.net.URLDecoder.decode(s, "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            return s;
        }
    }

    private static int getDefaultPort(String scheme) {
        return switch (scheme) {
            case "http" -> 80;
            case "https" -> 443;
            case "ftp" -> 21;
            case "ssh" -> 22;
            default -> -1;
        };
    }

    public String reconstruct() {
        StringBuilder sb = new StringBuilder();
        sb.append(scheme).append(":");
        if (host != null) {
            sb.append("//");
            if (userInfo != null) {
                sb.append(userInfo).append("@");
            }
            if (host.contains(":") && !host.startsWith("[")) {
                sb.append("[").append(host).append("]");
            } else {
                sb.append(host);
            }
            if (port != getDefaultPort(scheme) && port != -1) {
                sb.append(":").append(port);
            }
        }
        if (path != null && !path.isEmpty()) {
            sb.append(path);
        }
        if (query != null && !query.isEmpty()) {
            sb.append("?").append(query);
        }
        if (fragment != null && !fragment.isEmpty()) {
            sb.append("#").append(fragment);
        }
        return sb.toString();
    }

    public boolean isValid() {
        return scheme != null && !scheme.isEmpty();
    }

    public String getScheme() { return scheme; }
    public String getHost() { return host; }
    public int getPort() { return port; }
    public String getPath() { return path; }
    public String getQuery() { return query; }
    public String getFragment() { return fragment; }
    public String getUserInfo() { return userInfo; }
    public Map<String, String> getQueryParams() { return queryParams; }
    public String getOriginal() { return original; }

    private static class Builder {
        final String original;
        String scheme;
        String userInfo;
        String host;
        int port = -1;
        String path;
        String query;
        String fragment;
        Map<String, String> queryParams = Collections.emptyMap();

        Builder(String original) { this.original = original; }
        UrlParser build() { return new UrlParser(this); }
    }
}
